File: Clase semana 05/02 - Clases grabadas/2024/1C/Clase05-Miercoles/src/Complejo.cpp
/*
 * Complejo.cpp
 *
 *  Created on: 03/04/2024
 *      Author: algo2
 */

#include "Complejo.h"
#include <iostream>

Complejo::Complejo() {
	cantidadDeComplejos++;
	this->parteReal = 0;
	this->parteImaginaria = 0;
}

Complejo::Complejo(Complejo * complejo) {
	cantidadDeComplejos++;
	validar(complejo);
	this->parteReal = complejo->parteReal;
	this->parteImaginaria = complejo->parteImaginaria;
}

Complejo::Complejo(double parteReal, double parteImaginaria) {
	cantidadDeComplejos++;
	this->parteReal += parteReal;
	this->setParteImaginaria(parteImaginaria);
}

Complejo::~Complejo() {
}

double Complejo::getParteImaginaria() {
	return parteImaginaria;
}

void Complejo::setParteImaginaria(double imaginaria) {
	this->parteImaginaria = imaginaria;
}

double Complejo::getParteReal() {
	return parteReal;
}

void Complejo::setParteReal(double parteReal) {
	this->parteReal = parteReal;
}

void Complejo::validar(Complejo * complejo) {
	if (complejo == NULL) {
		throw "El complejo esta vacio";
	}
}

Complejo * Complejo::sumar(Complejo * complejo) {
	validar(complejo);
	Complejo * resultado = new Complejo();
	resultado->parteReal = this->parteReal + complejo->parteReal;
	resultado->parteImaginaria = this->parteImaginaria + complejo->parteImaginaria;
	return resultado;
}

Complejo * Complejo::duplicar() {
	Complejo * resultado = new Complejo();
	resultado->parteReal = this->parteReal;
	resultado->parteImaginaria = this->parteImaginaria;
	return resultado;

}

void Complejo::sumarse(Complejo * complejo)  {
	this->parteReal += complejo->parteReal;
	this->parteImaginaria += complejo->parteImaginaria;
}

std::string Complejo::imprimir() {
	std::string resultado = "Parte real: ";
	resultado += this->getParteReal();
	return resultado;
}

Complejo * Complejo::sumarComplejos(Complejo * termino1, Complejo * termino2) {
	Complejo * resultado = new Complejo();
	resultado->parteReal = termino1->parteReal + termino2->parteReal;
	resultado->parteImaginaria = termino1->parteImaginaria + termino2->parteImaginaria;
	return resultado;
}

File: Clase semana 07/02 - Clases grabadas/2024/2C/src/ar/uba/fi/cb100/semana03/miercoles/Complejo.java
package ar.uba.fi.cb100.semana03.miercoles;

import ar.uba.fi.cb100.semana04.miercoles.Numero;

public class Complejo extends Numero {
	
	private double parteImaginaria;

	public static Complejo construir(double parteReal, double parteImaginaria) {
		return new Complejo(parteReal, parteImaginaria);
	}
	
	private Complejo() {
		this(0.0, 0.0);
	}

	private Complejo(String parteReal) {
		this(Double.valueOf(parteReal), 0.0);
	}

	/**
	 * pre - 
	 * @param parteReal: cualquier valor flotante
	 * @param parteImaginaria: cualquier valor flotante
	 * 
	 * pos: crea el complejo con la partereal e imaginaria dadas
	 */
	private Complejo(double parteReal, double parteImaginaria) {		
		super(parteReal);
		this.parteImaginaria = parteImaginaria;
	}
	
	/**
	 * pre: -
	 * pos: suma la parte real dada por parametro al objeto actual
	 * @param parteReal valor entera
	 * 
	 * interface: public void sumarParteReal(int parteReal)
	 * nombre del metodo: sumarParteReal
	 * firma: sumarParteReal(int) (una sola por clase)
	 */
	public void sumarParteReal(int parteReal) {
		this.sumarParteReal( parteReal );
	}
	
	public void sumarParteReal(String parteReal) {
		this.sumarParteReal( Double.valueOf(parteReal) );
	}
	
	/**
	 * pre: -
	 * pos: suma la parte real dada por parametro al objeto actual
	 * @param parteReal valor entera
	 * 
	 * interface: public void sumarParteReal(int[] parteReal)
	 * nombre del metodo: sumarParteReal
	 * firma: sumarParteReal(int[]) (una sola por clase)
	 */
	public void sumarParteReal(int ... partesReales) {
		for(int i = 0; i < partesReales.length; i++) {
			this.sumarParteReal( partesReales[i] );
		}
	}
	
	/**
	 * pre: -
	 * post: imprime el complejo
	 */
	@Override
	public void imprimir() {
		System.out.println("El numero complejo es: " + this.getParteReal() + " + " + this.getParteImaginaria() + "i");
	}
	
	public void sumarParteReal(Complejo ... complejos) {
		for(int i = 0; i < complejos.length; i++) {
			this.sumarParteReal( complejos[i] );
		}
	}
	
	public double getParteImaginaria() {
		return parteImaginaria;
	}
	
	public void setParteImaginaria(double parteImaginaria) {
		this.parteImaginaria = parteImaginaria;
	}
	
	/**
	 * pre: complejo no puede ser nulo
	 * post: va a sumar en un nuevo complejo, el complejo actual y complejo por parametro.
	 * @param complejo: valor a sumar
	 * @return
	 * @throws Exception 
	 */
	public Complejo sumar(Complejo complejo) throws Exception {
		if (complejo == null) {
			throw new Exception("El parametro no puede ser nulo");
		}
		return new Complejo(this.getParteReal() + complejo.getParteReal(), this.parteImaginaria + complejo.getParteImaginaria());
	}
	
	@Override
	public void sumar(Numero numero) {
		//hacer la suma		
	}
	
	/**
	 * pre: no hay  
	 * @param parteReal valor a sumar
	 * @return devuelve la suma del complejo actual mas la parte real. La parte imaginaria queda como esta. 
	 */
	public Complejo sumar(double parteReal) {
		return new Complejo(this.getParteReal() + parteReal, this.parteImaginaria);
	}

}
